### ISSUE TITLE
"make check" unittests fail on Windows

### ISSUE THREAD

**Issue by eht16** [geany/geany#2677]
When running the unittests on Windows, all Ctags tests fail because TagManager writes the tag files with CRLF line endings but the tests compare against the test samples which have LF line endings (most of them).
There is a probably easy solution: as @techee pointed out in https://github.com/geany/geany/pull/2666#issuecomment-734731377, writing the tag file in binary mode will use LF line endings when writing the tags file (and *should* have no effect on non-Windows systems :crossed_fingers:).
There are six test samples which have CRLF line endings which still causes the tests to fail. Those could be converted to LF line endings:
- 3184782.sql.tags 
- 3526726.tex.tags 
- bug1570779.sql.tags 
- ingres_procedures.sql.tags 
- matlab_backtracking.m.tags 
- matlab_test.m.tags 

Apart from the Ctags unittests, we have some tests for various utility functions in https://github.com/geany/geany/blob/master/tests/test_utils.c where few filesystem related tests fail on Windows. Those could be probably `ifdef`'d for Windows and/or separate tests for Windows be added.

Once these issues are fixed, we could enable `make check` in CI!?

**Comment by b4n** [comment]
> There are six test samples which have CRLF line endings which still causes the tests to fail. Those could be converted to LF line endings:
> 
>     * 3184782.sql.tags
> 
>     * 3526726.tex.tags
> 
>     * bug1570779.sql.tags
> 
>     * ingres_procedures.sql.tags
> 
>     * matlab_backtracking.m.tags
> 
>     * matlab_test.m.tags

That' super duper weird, it's exactly the ones that would fail would CTags not normalize line ends itself (which it does).  If I manually remove that piece of code, I get those failures on Linux:
```diff
diff --git a/ctags/main/read.c b/ctags/main/read.c
index e19f15697..3586a1aaa 100644
--- a/ctags/main/read.c
+++ b/ctags/main/read.c
@@ -857,7 +857,7 @@ static eolType readLine (vString *const vLine, MIO *const mio)
          * used forms of line breaks are: LF (UNIX/Mac OS X), CR-LF (MS-DOS) and
          * CR (Mac OS 9). As CR-only EOL isn't handled by gets() and Mac OS 9
          * is dead, we only handle CR-LF EOLs and convert them into LF. */
-        if (newLine && vStringLength (vLine) > 1 &&
+        if (false && newLine && vStringLength (vLine) > 1 &&
             vStringChar (vLine, vStringLength (vLine) - 2) == '\r')
         {
             vStringChar (vLine, vStringLength (vLine) - 2) = '\n';
```

I really don't understand how Windows would make any difference here, I don't see any code being conditional around there.  And even if the libc was doing the conversion CRLF→LF behind the scenes, that should not be a problem either (unless maybe if there's a bug in the libc regarding fgetpos/fsetpos after such mappings, but I wouldn't expect this kind of breakage).

@eht16 Could you check what exactly is the return value of the `mio_gets()` calls in `readLine()` (or `iFileReadLine()` before *ctags_sync5*)?  The only explanation I have is that it doesn't return the format expected by this code -- which would be surprising, but well.

**Comment by techee** [comment]
I have a suspect. I haven't checked exactly how we read the files for unit tests but it could be that we use MIO_FILE so the underlying OS implementation is taken. Microsoft documentation for fgets()

https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/fgets-fgetws?view=msvc-160

says

```
...fgets reads characters from the current stream position to and including the first newline character...
```

and for `CRLF` line endings this means we could end with `CR` at the end of the line and then reading separate `LF`. (If this really is the case, it's just a purely idiotic way to implement this function.)

**Comment by b4n** [comment]
I really can't imagine Microsoft would have done that, as not only it's not conformant (newline character is `LF`) but it would also be super annoying on users (reading 2 "lines" for each one ending in `CRLF`).  But well, who knows, maybe in binary mode they do just that…?  Again, looking at the exact string returned by the `mio_gets()` call could be informative.

**Comment by techee** [comment]
Yeah, but the curious thing is that they explicitly mention the first newline character and don't say something like until newline like in the unix documentation of the function. But yeah, we'll know more when we get the output from Enrico.

**Comment by elextr** [comment]
I think they mean `newline` ie the LF character, thats what the C standard says and MS claims compatibility.

**Comment by eht16** [comment]
I still think reading the source files is not the problem. Tag parsing works fine on Windows, for LF and CRLF source files.

Anyway, I added some code in https://github.com/geany/geany/blob/c5303c9499bb6f0d632e0ca5b42e15f84d64ac52/ctags/main/read.c#L845 to print the return value of `mio_gets()`.

The source file content I used for testing, saved with CRLF line endings:
static void test_function(char *const str)
{
	printf("%s", str);
}

The hex representation of the `mio_gets()` result after the file is *opened from disk* (spaces and line breaks added by me):
73 74 61 74 69 63 20 76 6f 69 64 20 74 65 73 74 5f 66 75 6e 63 74 69 6f 6e 28 63 68 61 72 20 2a 63 6f 6e 73 74 20 73 74 72 29 0d 0a
7b 0d 0a
09 70 72 69 6e 74 66 28 22 25 73 22 2c 20 73 74 72 29 3b 0d 0a
7d 0d 0a

The hex representation of the `mio_gets()` result after the file was edited in Geany (spaces and line breaks added by me):
73 74 61 74 69 63 20 76 6f 69 64 20 74 65 73 74 5f 66 75 6e 63 74 69 6f 6e 28 63 68 61 72 20 2a 63 6f 6e 73 74 20 73 74 72 29 0d 0a
7b 0d 0a
09 70 72 69 6e 74 66 28 22 25 73 22 2c 20 73 74 72 29 3b 0d 0a
7d 0d 0a

As you can see, the read lines contain the CRLFs.

**Comment by techee** [comment]
> I still think reading the source files is not the problem. Tag parsing works fine on Windows, for LF and CRLF source files.

@eht16 I think we don't understand each other. We are trying to understand the results you got here:

https://gist.github.com/eht16/895090eddc23c6d241eb2cbec25159c0#file-test-suite-log-L91

For instance, you get this failure there

```
FAIL: ingres_procedures.sql
===========================

--- ./ingres_procedures.sql.tags	2016-06-12 17:16:40.683765400 +0200
+++ /tmp/tmp.2xGTegTO88/test.sql.tags	2020-11-25 23:33:27.945656500 +0100
@@ -1,7 +1,3 @@
 # format=tagmanager
 db0001Ì256Ö0
-db0002Ì256Ö0
-db0003Ì256Ö0
-errÌ16384Ö0
-nÌ16384Ö0
 xÌ16384Ö0
FAIL ingres_procedures.sql.tags (exit status: 2)
```

and the test fails not because of different line endings in the tags file, but because some real tags are missing in the output. From what I understood, when you change line endings from CRLF to LF in the test, this diff goes away. But then it means that the CRLFs in the test are the cause of the missing tags and there's some problem with parsing.

**Comment by techee** [comment]
@eht16 Oh and one more important thing - I think there's a difference when you run a unit test and when you simply open a file in Geany. When you open a file in Geany, ctags parses the buffer from Scintilla we pass to it using MIO_TYPE_MEMORY. But when you run a unit test, I think ctags parses the file using MIO_TYPE_FILE (I haven't actually checked it but I think this is the case). With MIO_TYPE_MEMORY you get completely platform-independent code because it just runs C code that simulates file operations. But when you use MIO_TYPE_FILE, the MIO library delegates the file operations to the corresponding OS calls and the result can differ based on what platform you use and I believe this could be our case.

**Comment by eht16** [comment]
> From what I understood, when you change line endings from CRLF to LF in the test, this diff goes away. But then it means that the CRLFs in the test are the cause of the missing tags and there's some problem with parsing.

You are completely right.
I was somewhat confused because the generated tags within Geany were always fine regardless of the line endings of the source file.
But when generating a tags file from the command line, I can see clearly the difference (as in the testsuite diff).

> @eht16 Oh and one more important thing - I think there's a difference when you run a unit test and when you simply open a file in Geany. When you open a file in Geany, ctags parses the buffer from Scintilla we pass to it using MIO_TYPE_MEMORY. But when you run a unit test, I think ctags parses the file using MIO_TYPE_FILE (I haven't actually checked it but I think this is the case). With MIO_TYPE_MEMORY you get completely platform-independent code because it just runs C code that simulates file operations. But when you use MIO_TYPE_FILE, the MIO library delegates the file operations to the corresponding OS calls and the result can differ based on what platform you use and I believe this could be our case.

I thought we use MIO_TYPE_FILE for the initial read when opening a file in Geany. But we do not.
So, yes, my assumption was wrong.

https://gist.github.com/eht16/998aae6eaf0da2cc2b45f998bf05606e this is `tests/ctags/3184782.sql`.
This is actually interesting as there seem to be CRCRLF line endings, maybe this is part of the problem.

It also seems that not just SQL, Matlab and TeX parsers are affected, to verify I changed one of the C test files to CRLF line endings and it failed too.

I won't have access to the my Windows machine for the next weeks (and much time at all). But I was able to reproduce the behavior when cross-compiling in a Docker container which makes it more easy to debug and it is way faster :). I'll try to share the Dockerfile and build script soon.

**Comment by b4n** [comment]
> This is actually interesting as there seem to be CRCRLF line endings, maybe this is part of the problem.

Hum :thinking: Could that be Git trying to be smart and convert LF to CRLF or something like that?

**Comment by elextr** [comment]
Maybe Geany's repository should gave a `.gitattributes` setting line endings in any case, so we don't accidently commit `crlf` from a windows user's PR.  The repository settings override the user `core.autocrlf` setting (or lack of it).

**Comment by b4n** [comment]
@elextr maybe yeah I never played with those so I'm not sure how it goes, but for the tests/ctags directory we definitely don't want *any* kind of conversion, we just want the same exact set of bytes.  So maybe we should tell Git to leave them alone and not be "smart" if that's the issue.

@eht16 could you check the md5/sha1 of the offending files on Windows and Linux? Or directly check the gitattribute idea if it makes sense.

**Comment by elextr** [comment]
@b4n yeah, my quick read of the docs (what!?!) suggests that .gitattributes can use globs to match files to have different conversion rules so should be able to force LF for code and binary for others.

**Comment by codebrainz** [comment]
FWIW, the Git installer for Windows asks you about how to deal with with line endings  (autocrlf) during setup and sets it globally. Maybe someone had that configured and accidentally changed the line endings?

![autocrlf](https://user-images.githubusercontent.com/181177/102153725-b751b300-3e2c-11eb-87be-21a4ed86703a.png)

**Comment by elextr** [comment]
The gitattributes now has an issue #2691 so we don't keep hijacking this one.
